agent/structs/config_entry_discoverychain_test.go

```diff
@@ -409,6 +409,636 @@ func TestConfigEntries_ListRelatedServices_AndACLs(t *testing.T) {
+				},
+				{
+					name:       "service read and mesh deny",
+					authorizer: newServiceAndMeshACL(t, "read", "deny"),
+					canRead:    true,
+					canWrite:   false,
+				},
+				{
+					name:       "service write and mesh deny",
+					authorizer: newServiceAndMeshACL(t, "write", "deny"),
+					canRead:    true,
+					canWrite:   false,
+				},
+
+				{
+					name:       "service deny and operator read",
+					authorizer: newServiceAndOperatorACL(t, "deny", "read"),
+					canRead:    false,
+					canWrite:   false,
+				},
+				{
+					name:       "service read and operator read",
+					authorizer: newServiceAndOperatorACL(t, "read", "read"),
+					canRead:    true,
+					canWrite:   false,
+				},
+				{
+					name:       "service write and operator read",
+					authorizer: newServiceAndOperatorACL(t, "write", "read"),
+					canRead:    true,
+					canWrite:   false,
+				},
+
+				{
+					name:       "service deny and operator write",
+					authorizer: newServiceAndOperatorACL(t, "deny", "write"),
+					canRead:    false,
+					canWrite:   true,
+				},
+				{
+					name:       "service read and operator write",
+					authorizer: newServiceAndOperatorACL(t, "read", "write"),
+					canRead:    true,
+					canWrite:   true,
+				},
+				{
+					name:       "service write and operator write",
+					authorizer: newServiceAndOperatorACL(t, "write", "write"),
+					canRead:    true,
+					canWrite:   true,
+				},
+
+				{
+					name:       "service deny and mesh read",
+					authorizer: newServiceAndMeshACL(t, "deny", "read"),
+					canRead:    false,
+					canWrite:   false,
+				},
+				{
+					name:       "service read and mesh read",
+					authorizer: newServiceAndMeshACL(t, "read", "read"),
+					canRead:    true,
+					canWrite:   false,
+				},
+				{
+					name:       "service write and mesh read",
+					authorizer: newServiceAndMeshACL(t, "write", "read"),
+					canRead:    true,
+					canWrite:   false,
+				},
+
+				{
+					name:       "service deny and mesh write",
+					authorizer: newServiceAndMeshACL(t, "deny", "write"),
+					canRead:    false,
+					canWrite:   true,
+				},
+				{
+					name:       "service read and mesh write",
+					authorizer: newServiceAndMeshACL(t, "read", "write"),
+					canRead:    true,
+					canWrite:   true,
+				},
+				{
+					name:       "service write and mesh write",
+					authorizer: newServiceAndMeshACL(t, "write", "write"),
+					canRead:    true,
+					canWrite:   true,
+				},
+			},
+		},
+		{
+			name:  "inline-certificate",
+			entry: &InlineCertificateConfigEntry{Name: "test"},
+			expectACLs: []testACL{
+				{
+					name:       "no-authz",
+					authorizer: newAuthz(t, ``),
+					canRead:    false,
+					canWrite:   false,
+				},
+
+				{
+					name:       "service deny and operator deny",
+					authorizer: newServiceAndOperatorACL(t, "deny", "deny"),
+					canRead:    false,
+					canWrite:   false,
+				},
+				{
+					name:       "service read and operator deny",
```

Comment:
So, are these attempts to test all possible combinations of where you can specify a `read`/`write` with some given name? I'm thinking we can probably drop the majority of these as relate to `service`-level permissions (since really only the APIGateway/BoundGateway config entries are related to given service instances) and just test mesh/operator-related permission combinations.

Reply:
> So, are these attempts to test all possible combinations of where you can specify a read/write with some given name?

Yes, these tests are slotting in matching existing patterns for other config entries